### ISSUE TITLE
[Q]: Security of controlled inputs for secrets 

### ISSUE THREAD

**Issue by kujon** [facebook/react#8736]
Hi,

I just run into an issue with controlled password fields when adding user tracking software to my app. Generally speaking, such trackers are going to claim that they don't capture text inputs. They achieve that by simply ignoring values of `change` events, they'll still however record the fact that the event has occurred, together with a selector uniquely identifying it's target.

When an input is controlled, the value attribute in the markup is going to change, resulting in a tracking software capturing it as part of the selector (`.some-input[type="password"][value="mypassword"]`), which brings me to my question: what are the security implications of having the value accessible not only via `HTMLInputElement.value`, but also by the virtue of parsing the markup?

Thanks,
K

**Comment by aweary** [comment]
If the value is accessible via the `value` property, then its already insecure as far as malicious access goes, right? Whether the value is set on the node, or as an attribute on the node, if one can be accessed so can the other.

The security concern, in this case, arises from the fact that your tracking software is sending arbitrary DOM markup over the network. Hypothetically, they could have just as easily created a unique selector based on the properties of the node instead of attributes (`.name`, `.type`, etc.) so it's not really a matter of whether one is less safe than the other: both are insecure and could contain sensitive user information.

This specific case could be mitigated by the user tracking software obscuring or ignoring the `value` attribute when `type="password"`. Using the value attribute isn't really a great way to create a unique selector either, especially with React apps since its updated every time a controlled input is updated. A workaround you could use for now would be to use uncontrolled components, since they don't update the `value` attribute on change.

**tl;dr**: if a program can execute code on the client, all local data is insecure, no matter what level obfuscation you implement. It's important that all executing programs carefully consider what kind of local data they send over the network, and that they do so securely.

**Comment by kujon** [comment]
@aweary thanks a lot for the comment! I've used the example of tracking software just for demonstration purposes - sure, if something can run in the same environment as my code, everything is fully compromised. My question is mostly about the fact that if the markup changes, I get an additional way of retrieving input's value and I'm wondering whether it can be exploited in some way.

**Comment by aweary** [comment]
I think it's safe to assume that if the markup is accessible to malicious attackers, everything else is. Avoiding this in the case of inputs sounds beneficial on the surface, but the added complexity wouldn't be worth the very minimal (if any) security benefits.

I'm going to go ahead and close this since there isn't anything actionable for the React team, but feel free to ask any clarifying questions!
